staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go

```diff
@@ -443,7 +443,29 @@ func (o *ApplyOptions) GetObjects() ([]*resource.Info, error) {
 			LabelSelectorParam(o.Selector).
 			Flatten().
 			Do()
+
 		o.objects, err = r.Infos()
+
+		if o.ApplySet != nil {
```

Comment:
Observations/suggestions:
- I was looking at the `o.Builder...` code earlier today for another PR and noticed that `Builder.Do` in particular uses a decorator pattern that is documented as being to provide "the opportunity to mutate the Info object or terminate early with an error." Might it make sense for the builder to have something like `.RequiredLabels(map[string]string)` that ends up applying those labels as part of the DecoratedVisitor call in `.Do`?
- If not, nit: I suggest extracting a function with the applyset portion of the code
- This happens ahead of the last-applied-configuration annotation calculation IIUC, so our label will show up as part of what the desired state it encodes. I think that's consistent with how Namespace is enforced (which seems to happen via `.Do` as I was describing above), but I just wanted to point that out for consideration.

Reply:
1) Opened https://github.com/kubernetes/kubernetes/pull/116069/files as an alternative, I think that I prefer this form because the builder logic is already pretty complex, and we probably want to retain more control over how we treat conflicting labels, rather than just have a generic error message.  That said, if we have other uses, then we should use the builder.

2) Split it out - good call!

3) Yes, I think that's right.  I think of the last-applied annotation as a simpler form of SSA field ownership.  This apply operation certainly owns the label on those objects.  For SSA we have the option of splitting it, but I don't think we want to do that, if anything it helps us understand that this apply operation is the "owning apply".

Reply:
1. It's a good question whether or not it could be reused--it's hard to say, but looking around the code, `kubectl label` seems like the main candidate, and it would need more complex behaviour. I think the new method-based implementation you did for (2) is also very clean, and I'm happy to stick with that.